### PR TITLE
Add transport support for dnsaddr

### PR DESCRIPTION
**Pull request by ackintosh**
Related issue: https://github.com/libp2p/rust-libp2p/issues/967
Related PR: https://github.com/libp2p/rust-libp2p/pull/1356

https://github.com/multiformats/multiaddr/blob/master/protocols/DNSADDR.md

### CODE REVIEW

File: transports/dns/src/lib.rs [hunk 1]
```diff
@@ -170,7 +176,56 @@ where
                             })
                             .next()
                             .ok_or_else(|| DnsErr::ResolveFail(name))
-                    }.left_future()
+                    }.boxed().left_future()
```
Comment (tomaka) [inline transports/dns/src/lib.rs]: You normally don't have to box the future. Just put an `async` block around the entire code instead of having two distinct `async` blocks.

Since this is a lot of code, it might be worth moving it to a separate freestanding function.

File: transports/dns/src/lib.rs [hunk 2]
```diff
@@ -299,10 +370,59 @@ mod tests {
...
+            fn listen_on(self, _: Multiaddr) -> Result<Self::Listener, TransportError<Self::Error>> {
+                unreachable!()
+            }
+
+            fn dial(self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>> {
+                let addr = addr.iter().collect::<Vec<_>>();
+                assert_eq!(addr.len(), 3);
+                // TODO: ipfs
+//                match addr[2] {
+//                    Protocol::Ipfs(_) => (),
```
Comment (ackintosh) [inline transports/dns/src/lib.rs]: I'll file another PR to add the ipfs support.

File: transports/dns/src/lib.rs [hunk 3]
```diff
@@ -238,6 +225,90 @@ where TErr: error::Error + 'static
...
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            // if resolved_addrs.len() == 0 {} // TODO
+
+            if suffix.len() == 0 {
+                Ok(resolved_addrs[0].clone())
+            } else {
+                // TODO: suffix matching
```
Comment (tomaka) [inline transports/dns/src/lib.rs]: I'd that this is out of scope of this function.

File: transports/dns/src/lib.rs [hunk 4]
```diff
@@ -238,6 +225,90 @@ where TErr: error::Error + 'static
...
+                    } else {
+                        None
+                    }
+                })
+                .next()
+                .ok_or_else(|| DnsErr::ResolveFail(name))
+        }
+        Protocol::Dnsaddr(ref n) => {
+            let conn = UdpClientConnection::new("8.8.8.8:53".parse().unwrap()).unwrap(); // TODO: error handling
+            let client = SyncClient::new(conn); // TODO: should use async client?
```
Comment (tomaka) [inline transports/dns/src/lib.rs]: > // TODO: should use async client?

Yes!

File: transports/dns/src/lib.rs (hunk 2 above)
Comment (ackintosh) [reply transports/dns/src/lib.rs]: https://github.com/libp2p/rust-libp2p/pull/1463

File: transports/dns/src/lib.rs [hunk 5]
```diff
@@ -238,6 +225,90 @@ where TErr: error::Error + 'static
     }
 }
 
+async fn resolve_dns<T>(p: Protocol<'_>, suffix: Multiaddr) -> Result<Multiaddr, DnsErr<T::Error>>
+    where
```
[the remark this answers is not in the document]
Comment (ackintosh) [reply transports/dns/src/lib.rs]: Updated at https://github.com/libp2p/rust-libp2p/pull/1450/commits/fcc3d6a0f9dd1d0e0d2d6245c56f41bbd882d0cc .

If `T: Transport` is removed we'll get the error:
```text
error[E0220]: associated type `Error` not found for `T`
   --> transports/dns/src/lib.rs:228:89
    |
228 | async fn resolve_dns<T>(p: Protocol<'_>, suffix: Multiaddr) -> Result<Multiaddr, DnsErr<T::Error>>
    |                                                                                         ^^^^^^^^ associated type `Error` not found
```

File: transports/dns/src/lib.rs (hunk 4 above)
Comment (ackintosh) [reply transports/dns/src/lib.rs]: Hmm... 

I've added [trust-dns](https://github.com/bluejekyll/trust-dns) in order to query TXT records. I'm trying to switch the SyncClient to AsyncClient but I think AsyncClient can not be used for rust-libp2p because a background future is required for the client. 🤔

https://docs.rs/trust-dns-client/0.19.3/trust_dns_client/#async-usage

```rust
// Create a new client, the bg is a background future which handles
//   the multiplexing of the DNS requests to the server.
//   the client is a handle to an unbounded queue for sending requests via the
//   background. The background must be scheduled to run before the client can
//   send any dns requests
let client = AsyncClient::connect(stream);

// await the connection to be established
let (mut client, bg) = runtime.block_on(client).expect("connection failed");

// make sure to run the background task
runtime.spawn(bg);
```

======

Do you know other library what can query TXT records asynchronously? 💦

Comment (tomaka) [reply transports/dns/src/lib.rs]: > I think AsyncClient can not be used for rust-libp2p because a background future is required for the client.

That makes it more difficult to implement, but normally not impossible.

Comment (tomaka) [reply transports/dns/src/lib.rs]: The idea is not to run the background future in the background, but to run it alongside with the "front" one. I don't have more precise guidelines without looking deeply into details (which I don't really have the time right now to do) though.

Comment (ackintosh) [reply transports/dns/src/lib.rs]: I've thought for a while about this but I can not come up anything to address the issue... hmm... 🤔💦

**Comment by romanb** [comment]
Superseded by https://github.com/libp2p/rust-libp2p/pull/1931.
